### PR TITLE
Dilithium: update to v3.1

### LINKED ISSUE #160: Update Dilithium to Round 3
Specification is expected to be released within two weeks.

### PR DESCRIPTION
**Pull request by bwesterb**
This updates Dilithium to v3.1 as described in

  https://pq-crystals.org/dilithium/data/dilithium-specification-round3-20210208.pdf

Some AVX2 optimisations have not yet been upgraded.

Change the hybrid schemes "eddilithium" accordingly.

Closes #160

### CODE REVIEW

**Review by claucece — APPROVED** [review]
LGTM <3

File: sign/dilithium/mode2/internal/sample.go [hunk 1]
```diff
@@ -233,226 +140,131 @@ func PolyDeriveUniform(p *common.Poly, seed *[32]byte, nonce uint16) {
...
-				t2 = uint32(buf[j]) >> 4
-			}
-			if t1 <= 2*Eta {
-				p[i] = common.Q + Eta - t1
-				i++
-			}
-			if t2 <= 2*Eta && i < common.N {
-				p[i] = common.Q + Eta - t2
-				i++
+				panic("unsupported η")
```
Comment (armfazh) [inline sign/dilithium/mode2/internal/sample.go:180]: unicode to the terminal!

File: sign/dilithium/mode2/internal/sample_test.go [hunk 2]
```diff
@@ -138,119 +138,24 @@ func TestDeriveUniformLeqEta(t *testing.T) {
...
 func TestDeriveUniformLeGamma1(t *testing.T) {
 	var p common.Poly
-	var seed [48]byte
+	var seed [64]byte
 	for i := 0; i < 100; i++ {
 		binary.LittleEndian.PutUint64(seed[:], uint64(i))
 		PolyDeriveUniformLeGamma1(&p, &seed, uint16(i))
 		for j := 0; j < common.N; j++ {
-			if p[j] < common.Q-common.Gamma1 || p[j] > common.Q+common.Gamma1 {
+			if p[j] > Gamma1 && p[j] <= common.Q-Gamma1 {
```
Comment (armfazh) [inline sign/dilithium/mode2/internal/sample_test.go:146]: Just to double check: 

For mode2:
```python
(p > 2^17) && (p  <= 8380417-2^17)
p> 131072  &&  p <=  8249345
```

then
```
-------(================]-------
     131072  ^fails^    8249345
```

File: sign/eddilithium3/eddilithium_test.go [hunk 3]
```diff
@@ -12,7 +12,7 @@ func BenchmarkVerify(b *testing.B) {
 	// instead of at the moment of verification (as compared to the reference
 	// implementation.  A fair comparison thus should sum verification
 	// times with unpacking times.)
-	var seed [32]byte
+	var seed [57]byte
```
Comment (armfazh) [inline sign/eddilithium3/eddilithium_test.go:15]: Seed is 32 byes for all modes, isn't it?

**Review by armfazh — COMMENTED** [review]
Few comments and some double-checks.

File: sign/dilithium/internal/common/asm/go.sum [hunk 4]
```diff
@@ -1,25 +1,55 @@
+github.com/bwesterb/go-ristretto v1.2.0 h1:xxWOVbN5m8NNKiSDZXE1jtZvZnC6JSJ9cYFADiZcWtw=
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/dilithium/internal/common/asm/go.sum:1]:
```
 $ rm go.sum
$ go mod download
$ go run src.go -out ../amd64.s -stubs ../stubs_amd64.go -pkg common
src.go:10:2: missing go.sum entry for module providing package github.com/mmcloughlin/avo/build; to add:
	go mod download github.com/mmcloughlin/avo
src.go:11:2: missing go.sum entry for module providing package github.com/mmcloughlin/avo/operand; to add:
	go mod download github.com/mmcloughlin/avo
src.go:12:2: missing go.sum entry for module providing package github.com/mmcloughlin/avo/reg; to add:
	go mod download github.com/mmcloughlin/avo
```

[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/dilithium/internal/common/asm/go.sum:1]: `go generat`, btw, is broken by this commit 3a9a68d7e373f25a0025c2b0c98d730274120c08

File: sign/dilithium/internal/common/field_test.go [hunk 5]
```diff
@@ -66,63 +63,8 @@ func TestPower2Round(t *testing.T) {
 		if (-(1 << (D - 1)) >= a0) || (a0 > 1<<(D-1)) {
 			t.Fatalf("power2round(%v): a0 out of bounds", a)
 		}
-		if a1 > (1 << (23 - 14)) { // 23 ~ 2log Q.
+		if a1 > (1 << (23 - 13)) { // 23 ~ 2log Q.
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/dilithium/internal/common/field_test.go:66]: Done.

File: sign/dilithium/mode2/internal/sample.go (hunk 1 above)
Comment (bwesterb) [reply sign/dilithium/mode2/internal/sample.go:180]: I don't see an issue with that.

File: sign/dilithium/mode2/internal/sample_test.go (hunk 2 above)
Comment (bwesterb) [reply sign/dilithium/mode2/internal/sample_test.go:146]: Yes, this is correct.  I might as well check that `p` is not equal to or larger than `q` which I added.

File: sign/eddilithium3/eddilithium_test.go (hunk 3 above)
Comment (bwesterb) [reply sign/eddilithium3/eddilithium_test.go:15]: Yes, but Ed448 requires a larger seed.

File: sign/dilithium/mode2/internal/sample.go (hunk 1 above)
Comment (armfazh) [reply sign/dilithium/mode2/internal/sample.go:180]: not all terminals will render unicode as expected.

File: sign/dilithium/internal/common/asm/go.sum (hunk 4 above)
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/dilithium/internal/common/asm/go.sum:1]: and what happens after running `go mod download github.com/mmcloughlin/avo` ?


> `go generate`, btw, is broken by this commit [3a9a68d](https://github.com/cloudflare/circl/commit/3a9a68d7e373f25a0025c2b0c98d730274120c08)

can this be fixed? or passing `-tags ignore` helps a bit?

File: sign/dilithium/mode2/internal/sample.go (hunk 1 above)
Comment (bwesterb) [reply sign/dilithium/mode2/internal/sample.go:180]: Go allows (certain) unicode characters in identifiers (eg `struct T食 { // ...`) and will print them happily to the terminal on compiler warnings/errors.  Many commonly used CLI tools (including our own `wrangler`) print unicode to the terminal. I think by now we can expect terminals to handle unicode properly just like we expect that from editors.

File: sign/dilithium/internal/common/asm/go.sum (hunk 4 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/dilithium/internal/common/asm/go.sum:1]: Fixed in 3543c1a54a659efb0da40225aef29ab3fa42f466

**Review by armfazh — APPROVED** [review]

**Pull request merged**